components/webgpu/lib.rs

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#[macro_use]
+extern crate serde;
+
+use ipc_channel::ipc::{self, IpcReceiver, IpcSender};
+use servo_config::pref;
+use std::sync::{Arc, Mutex};
+pub use wgpu_native as wgpu;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WebGPUResponse {
+    RequestAdapter(String, Arc<WebGPUAdapter>),
+    RequestDevice,
+}
+
+pub type WebGPUResponseResult = Result<WebGPUResponse, String>;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WebGPURequest {
+    RequestAdapter(IpcSender<WebGPUResponseResult>, wgpu::RequestAdapterOptions),
+    RequestDevice,
+    Exit,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct WebGPUThread(pub IpcSender<WebGPURequest>);
+
+impl WebGPUThread {
+    pub fn new() -> Option<WebGPUThread> {
+        if !pref!(dom.webgpu.enabled) {
+            return None;
+        }
+        let (sender, receiver) = ipc::channel().unwrap();
+
+        std::thread::Builder::new()
+            .name("WebGPUThread".to_owned())
+            .spawn(move || {
+                WebGPU::new(receiver).run();
+            })
+            .expect("Thread spawning failed");
+        Some(WebGPUThread(sender))
+    }
+
+    pub fn exit(&self) -> Result<(), &'static str> {
+        self.0
+            .send(WebGPURequest::Exit)
+            .map_err(|_| "Failed to send Exit message")
+    }
+}
+
+struct WebGPU {
+    receiver: IpcReceiver<WebGPURequest>,
+    adapters: Vec<Arc<WebGPUAdapter>>,
+}
+
+impl WebGPU {
+    fn new(receiver: IpcReceiver<WebGPURequest>) -> Self {
+        WebGPU {
+            receiver,
+            adapters: Vec::new(),
+        }
+    }
+
+    fn run(&mut self) {
+        while let Ok(msg) = self.receiver.recv() {
+            match msg {
+                WebGPURequest::RequestAdapter(sender, options) => {
+                    let id = wgpu::wgpu_request_adapter(Some(&options));
+                    let adapter = Arc::new(WebGPUAdapter {
+                        id,
+                        invalid: Mutex::new(false),
```

Comment:
I'm not sure what you plan to do with the `invalid` flag, and I can point out that storing the `Arc` here and then sending a clone over IPC will not result in sharing the same data between the gpu-thread and the script-thread.

Reply:
Thanks for catching this, I've changed the way we could identify invalid adapters in the future, by tracking the invalid IDs in the webgpu thread.